### ISSUE TITLE
Backtesting question: custom_trade_info / trade.get_custom_data(key) 

### ISSUE THREAD

**Issue by trackersNZ** [freqtrade/freqtrade#11764]
## Describe your environment

  * Operating system: Ubuntu 22.04
  * Python Version: 3.12.3
  * CCXT version:4.4.75
  * Freqtrade Version: 2025.4
  

**Describe the bug**
In backtesting mode, custom data configured via the custom_trade_info() method and set in the dataframe during populate_entry_trend is not accessible on the Trade object via trade.get_custom_data(key) within callback methods like custom_stoploss or custom_exit. The trade.get_custom_data(key) call consistently returns None, even for simple, hardcoded integer values.
This prevents dynamic, entry-specific data (like ATR at entry) from being used in custom stop-loss or take-profit logic during backtests.

**Steps To Reproduce**
Create a minimal strategy file named TestCustomInfo.py in user_data/strategies/ with the following content:

user_data/strategies/TestCustomInfo.py

``` python
from freqtrade.strategy import IStrategy
from pandas import DataFrame
import logging # Added for logger

logger = logging.getLogger(__name__) # Added for logger

class TestCustomInfo(IStrategy):
    timeframe = '4h'
    minimal_roi = {"0": 100}
    stoploss = -0.10 # Using a tighter, more realistic fallback for testing
    use_custom_stoploss = True

    # Minimal startup candles
    startup_candle_count: int = 1 

    def custom_trade_info(self) -> dict:
        return {'my_test_key': 'df_column_for_test'}

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # Ensure the column exists and is initialized for all rows before populate_entry_trend
        dataframe['df_column_for_test'] = 0 # Initialize with a default
        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        # Set the custom data value for the entry candle
        # For simplicity, let's just try to set it for the first few available candles
        if len(dataframe) > 1:
            dataframe.loc[dataframe.index[1], 'df_column_for_test'] = 123 
            dataframe.loc[dataframe.index[1], 'enter_long'] = 1
            dataframe.loc[dataframe.index[1], 'enter_tag'] = 'test_entry_candle_1'
        else:
            dataframe['enter_long'] = 0 # Ensure column exists
        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['exit_long'] = 0
        return dataframe
    
    def custom_stoploss(self, pair: str, trade, current_time, current_rate,
                        current_profit: float, **kwargs) -> float:
        
        test_value_sl = None
        try:
            test_value_sl = trade.get_custom_data('my_test_key')
        except Exception as e:
            logger.error(f"Error in custom_stoploss accessing 'my_test_key': {e}")

        logger.info(f"CUSTOM_STOPLOSS for {trade.id} ({pair}): 'my_test_key' = {test_value_sl} (type: {type(test_value_sl)})")
        
        if test_value_sl == 123:
            logger.info(f"SUCCESS in custom_stoploss: 'my_test_key' was correctly retrieved as 123!")
        else:
            logger.info(f"FAILURE in custom_stoploss: 'my_test_key' was NOT 123. Value: {test_value_sl}")
        
        # Exit immediately after one candle for test, using a very tight stop
        # to ensure this callback is triggered and the trade closes.
        return trade.open_rate * 0.99
```

Python
Ensure you have some 4h data for ETH/USDT:USDT (e.g., for 2023-01-01 to 2023-01-03).

Run the following backtesting command:
freqtrade backtesting --config user_data/config.json --strategy TestCustomInfo --timerange 20230101-20230103 --pairs ETH/USDT:USDT --stake-amount 100

(Adjust timerange and pairs if necessary, but ensure at least one trade occurs).

**Observed behavior**
The log output from custom_stoploss will show:
INFO - TestCustomInfo - CUSTOM_STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
INFO - TestCustomInfo - FAILURE in custom_stoploss: 'my_test_key' was NOT 123. Value: None

This indicates that trade.get_custom_data('my_test_key') returns None, even though the dataframe column df_column_for_test was set to 123 on the entry candle.

**Expected behavior**
trade.get_custom_data('my_test_key') within custom_stoploss (and other callbacks like custom_exit) should return the value (123 in this test case) that was set in the corresponding dataframe column (df_column_for_test) on the entry signal candle, as mapped by the custom_trade_info() method.

**Log output**

```
2025-05-18 00:25:12,456 - freqtrade.optimize.backtesting - INFO - Backtesting with data from 2023-01-01 00:00:00 up to 2023-01-02 00:00:00 (1 days).
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None
!!!!!!!! INSIDE CUSTOM STOPLOSS for 1 (ETH/USDT:USDT): 'my_test_key' = None (type: <class 'NoneType'>)
!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: None

...
... (repeated for each candle the trade is open) ...
```

Strategy file
user_data/strategies/TestCustomInfo.py
``` python
from freqtrade.strategy import IStrategy
from pandas import DataFrame

class TestCustomInfo(IStrategy):
    timeframe = '4h'
    minimal_roi = {"0": 100}
    stoploss = -0.99 # Will use custom stoploss for test

    use_custom_stoploss = True # To trigger custom_stoploss

    def custom_trade_info(self) -> dict:
        return {'my_test_key': 'df_column_for_test'}

    def populate_indicators(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['df_column_for_test'] = 123 # Constant value
        return dataframe

    def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['enter_long'] = 1 # Enter on every candle for testing
        dataframe['enter_tag'] = 'test_entry'
        return dataframe

    def populate_exit_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
        dataframe['exit_long'] = 0
        return dataframe
    
    def custom_stoploss(self, pair: str, trade, current_time, current_rate,
                        current_profit: float, **kwargs) -> float:
        test_value = trade.get_custom_data('my_test_key')
        print(f"!!!!!!!! INSIDE CUSTOM STOPLOSS for {trade.id} ({pair}): 'my_test_key' = {test_value} (type: {type(test_value)})")
        if test_value == 123:
            print(f"!!!!!!!! SUCCESS: 'my_test_key' was correctly retrieved as 123!")
        else:
            print(f"!!!!!!!! FAILURE: 'my_test_key' was NOT 123. Value: {test_value}")
        return -0.01 # Exit immediately after one candle for test
```

Config file snippet:
```
    "$schema": "https://schema.freqtrade.io/schema.json",
    "max_open_trades": 8,
    "stake_currency": "USDC",
    "stake_amount": "unlimited",
    "trading_mode": "futures", // Ensures futures mode
    "margin_mode": "isolated", // Or "cross"
    "leverage": 5,             // Ensure this is set if you want leverage
    "tradable_balance_ratio": 0.99,
    "fiat_display_currency": "NZD",
    "dry_run": false,
    "dry_run_wallet": 1000,
    "cancel_open_orders_on_exit": false,
	"liquidation_buffer": 0.05,

```

Additional context
This issue prevents the use of dynamic entry-specific values (like ATR at entry) for custom stop-loss or take-profit calculations during backtesting, as the custom_trade_info mechanism does not appear to be populating the Trade object's custom data store correctly. The feature works as expected in live/dry-run modes (if you have tested this part and can confirm).

**Comment by stash86** [comment]
Please stop using AI and read the docs properly.

This is the correct way of saving and retrieving custom data
https://www.freqtrade.io/en/stable/strategy-advanced/#storing-information-persistent

This is an example of how to store and use ATR for exit
https://botacademy.ddns.net/2024/04/18/fsd-better-way-of-using-atr-for-exits/

**Comment by xmatthias** [comment]
Well your strategy never sets anything. 
Just because your AI generates a random function doesn't mean freqtrade will care about it or call it.

**Used** callbacks are listed in the [callbacks documentation](https://www.freqtrade.io/en/stable/strategy-callbacks/).
Random functions you throw into your strategy are however not called - and that's also not planned.

PLEASE, PLEASE (every future reader, too) - **before** posting an issue here, make sure to read the documentation.
AI does NOT do that, but tries to guess something that looks roughly like it could work - but doesn't care about the documentation.

We're here to look at bugs, or help **real** people with their strategies - but trying to help your AI is NOT what we aim to spend our time on.

This is also mentioned pretty clearly in the Question issue template (the correct template if your strategy is not working).
![Image](https://github.com/user-attachments/assets/ac032f01-1b33-49a0-bd5a-1ea8538b7e2c)

I'll however add it to the bug report template, too - to be even clearer about this.

**Comment by trackersNZ** [comment]
I have learned a lot, thanks for the advice and apologies for wasting your time :)
